### DISCUSSION TITLE
LoRaWANNode.setDatarate() parameter

### DISCUSSION THREAD

**Discussion by ksjh** [jgromes/RadioLib#917] [Q&A] [answered]
I am trying to get a Heltec Wireless Tracker to work as a node for TTNMapper without using any Heltec library. Everything basically works, the RadioLib library is great. But I am not sure what parameter to use as an argument for the `LoRaWANNode.setDatarate()` function. Looking at the code, I suspect that it should be an index for the array `dataRates` of the respective `FrequencyBand`. Or does this function expect a bit field composed of the constants `RADIOLIB_LORAWAN_DATA_RATE_...` in `LoRaWAN.h`?

Due to some ADR stuff, my application currently uses `SF12BW125` for connecting to my own next door SX1301 Basics Station gateway. I have turned ADR off in my application (`node.setADR(false);`). Initially, `SF10BW125` was used, but after some testing in more remote locations, the data rate seems to be stuck at this value now. I am now trying to manually set a fixed data rate, but  so far, I had no success.

**Comment by jgromes** [comment, marked as answer]
`setDatarate` expects a LoRaWAN data rate "ID", which is just a number, 0 to 15. What values of spreading factor and bandwidth that data rate represents depends on the frequency band you are using. For example, in EU868 band, DR0 is SF12BW125, while in US915 it's SF10BW125.

> I have turned ADR off in my application

If you do that, then you will have to make sure that the end device, the gateway and the TTN server/application all use the same fixed data rate. If the TTN application is still attempting to use some old data rate obtained during ADR then the node will not be able to uplink. And since ADR is disabled, there is no mechanism to fallback to some default value.

**Reply by ksjh** [reply]
Thank you. As I understand it now, the expected parameter is essentially the **index** of the requested data rate in the `dateRate[]` array for the used `LoRaWANFrequency_t` from `src/protocols/LoRaWAN/LoRaWANBands.cpp`. Great. Let's see if I can make it work with this information.

**Reply by ksjh** [reply]
It worked as expected. I did not have to change anything on the gateway or app server. Once I initialized a new `node.beginOTAA()` where I also set the new fixed data rate as the `joinDR` and used the stored EEPROM data otherwise, my gateway immediately received the uplink data with the desired `SF7BW125`. The app server and integrations also received the data correctly after the update of my code on the node. No change needed server-side. My gateway is a multi-channel SX1301 gateway with the newer Semtech Basics Station software. It should adapt automatically. And it did. My current TTN application does not send anything in the downlink direction.

**Reply by jgromes** [reply]
> My current TTN application does not send anything in the downlink direction

That's not entirely true, the server will most likely be sending some MAC commands even if ADR is disabled. So you still need to make sure the downlink channel works.

**Reply by ksjh** [reply]
Thanks, that's why I wrote "my application". I am using the EU868-870 ISM band. As far as I understand table 9 of the regional parameter document 1.1, the data rate in the RX1 window should also adapt automatically, once the network recognizes the new uplink data rate. For RX2, it is fixed anyway. So, do I now need to set the data rate on the node by hand before calling `LoRaWANNode.downlink()`? I thought the change of the data rate for the RX windows was done on the node automatically, even when disabling ADR. I could be mistaken, but so far, I did not notice any errors or other negative effects without doing this manually.

**Reply by StevenCellist** [reply]
Correct, just calling `.setDatarate()` before sending an uplink does not need any TTN-side modifications nor further changes in your code.

**Reply by ksjh** [reply]
Thank you @StevenCellist, this is what I have been doing in my latest experiments, setting the data rate explicitly before each uplink. Before, I tried to disable ADR right after OTAA and setting the data rate only once. I thought this might keep the data rate constant. But after moving the node around for mapping, I noticed that the rate had changed to `SF12BW125` and stayed like that, even back at home right next to the gateway. I suspect that this might be due to the things done after the comment `// check if we need to do ADR stuff` in `src/protocols/LoRaWAN/LoRaWAN.cpp`. Briefly looking at this code, I get the idea that the data rate is only ever decreased here, never increased. Ist this true and intentional?

**Reply by StevenCellist** [reply]
@ksjh LOL good catch. That whole block of code should be guarded by an if-clause that checks whether ADR is actually enabled. Also explains one of the things I encountered recently with my Wireless Tracker-mapper. Adding to #918.
